src/Primitives/arduino.cpp

```diff
@@ -150,15 +150,32 @@ int prim_index = 0;
             PrimitiveEntry *p = &primitives[prim_index++];                 \
             p->name = #prim_name;                                          \
             p->f = &(prim_name);                                           \
+            p->f_reverse = nullptr;                                        \
+            p->f_serialize_state = nullptr;                                \
         } else {                                                           \
             FATAL("pim_index out of bounds");                              \
         }                                                                  \
     }
 
+#define install_primitive_reverse(prim_name)             \
+    {                                                    \
+        PrimitiveEntry *p = &primitives[prim_index - 1]; \
+        p->f_reverse = &(prim_name##_reverse);           \
+        p->f_serialize_state = &(prim_name##_serialize); \
+    }
+
 #define def_prim(function_name, type) \
     Type function_name##_type = type; \
     bool function_name(Module *m)
 
+#define def_prim_reverse(function_name)     \
+    void function_name##_reverse(Module *m, \
+                                 std::vector<IOStateElement> external_state)
+
+#define def_prim_serialize(function_name) \
+    void function_name##_serialize(       \
+        std::vector<IOStateElement *> &external_state)
+
```

Comment:
All the def and install primitives are the same across platforms. We should move this to `primitives.h`